packages/koa-shopify-auth/src/auth/index.ts

```diff
@@ -3,36 +3,77 @@ import {Context} from 'koa';
 import {OAuthStartOptions, AccessMode, NextFunction} from '../types';
 import createOAuthStart from './create-oauth-start';
 import createOAuthCallback from './create-oauth-callback';
+import createEnableCookies from './create-enable-cookies';
+import createTopLevelRedirect from './create-top-level-redirect';
 
+const DEFAULT_MYSHOPIFY_DOMAIN = 'myshopify.com';
 const DEFAULT_ACCESS_MODE: AccessMode = 'online';
 
+const TOP_LEVEL_OAUTH_COOKIE_NAME = 'shopifyTopLevelOAuth';
+export const TEST_COOKIE_NAME = 'shopifyTestCookie';
+
+function hasCookieAccess({cookies}: Context) {
+  return !!cookies.get(TEST_COOKIE_NAME);
+}
+
+function shouldPerformInlineOAuth({cookies}: Context) {
+  return !!cookies.get(TOP_LEVEL_OAUTH_COOKIE_NAME);
+}
+
 export default function createShopifyAuth(options: OAuthStartOptions) {
   const config = {
     scopes: [],
     prefix: '',
+    myShopifyDomain: DEFAULT_MYSHOPIFY_DOMAIN,
     accessMode: DEFAULT_ACCESS_MODE,
     ...options,
   };
 
   const {prefix} = config;
 
   const oAuthStartPath = `${prefix}/auth`;
-  const oAuthStart = createOAuthStart(config);
-
   const oAuthCallbackPath = `${oAuthStartPath}/callback`;
+
+  const oAuthStart = createOAuthStart(config, oAuthCallbackPath);
   const oAuthCallback = createOAuthCallback(config);
 
+  const inlineOAuthPath = `${prefix}/auth/inline`
+  const redirectToTopLevelOAuth = createTopLevelRedirect(inlineOAuthPath);
+
+  const enableCookiesPath = `${oAuthStartPath}/enable_cookies`;
+  const enableCookies = createEnableCookies(config);
+  const redirectToEnableCookies = createTopLevelRedirect(enableCookiesPath);
+
   return async function shopifyAuth(ctx: Context, next: NextFunction) {
-    if (ctx.path === oAuthStartPath) {
+    if (ctx.path === oAuthStartPath && !hasCookieAccess(ctx)) {
+      // This is to avoid a redirect loop if the app doesn't use verifyRequest or set the test cookie elsewhere.
+      ctx.cookies.set(TEST_COOKIE_NAME, '1');
+      await redirectToEnableCookies(ctx);
+      return;
+    }
+
+    if (ctx.path == inlineOAuthPath || (ctx.path === oAuthStartPath && shouldPerformInlineOAuth(ctx))) {
+      ctx.cookies.set(TOP_LEVEL_OAUTH_COOKIE_NAME);
       await oAuthStart(ctx);
       return;
     }
 
+    if (ctx.path === oAuthStartPath) {
+      ctx.cookies.set(TOP_LEVEL_OAUTH_COOKIE_NAME, '1');
```

Comment:
it feels kind of weird to set cookies in this function since otherwise it's only concern is routing.